File: books/cs/general-purpose-algorithms/introduction-to-algorithms/chapter8/Go/counting_sort.go
ckage main

import (
    "fmt"
        "time"
    "math/rand"
)

const dataset_size = 100
const dataset_max_value = 100

func main() {
	var aaa[dataset_size]int
	var bbb[dataset_size+1]int
	var ccc[dataset_max_value]int

	max_value := 0
	total := 0


	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	for idx:=0; idx<dataset_size; idx++ {
		aaa[idx] = r.Intn(dataset_max_value)
		if( max_value < aaa[idx] ){
			max_value = aaa[idx]
		}
	}
	fmt.Println(aaa)

	for idx:=0; idx<dataset_size; idx++ {
		ccc[aaa[idx]] += 1
	}
	for idx:=0; idx<=max_value; idx++ {
		total += ccc[idx]
		ccc[idx] = total
	}
	fmt.Println(ccc)
	for idx:=dataset_size-1; idx>=0; idx-- {
		bbb[ccc[aaa[idx]]] = aaa[idx]
		ccc[aaa[idx]] -= 1
	}
	fmt.Println(bbb)
}

File: books/cs/general-purpose-algorithms/introduction-to-algorithms/chapter8/PERL/counting_sort.pl
#!/usr/bin/perl
use List::Util qw(max);

my $dataset_size = 100;
my $random_max = 100;
my $k = 0;
my $total = 0;
my $idx = 0;

my @aaa = (0) x $dataset_size;
my @bbb = (0) x $dataset_size;
my @ccc = ();

for($idx=0;$idx<$dataset_size;$idx++){
	$aaa[$idx] = int(rand($random_max));
	print $aaa[$idx],"\n";
}

$k = max(@aaa);
@ccc = (0) x ($k+1);
for my $idx (@aaa) {
	$ccc[$idx] += 1;
}
print @ccc,"\n";
$total = 0;
for($idx=0; $idx<=$k; $idx++){
	$total += $ccc[$idx];
	$ccc[$idx] = $total;
}
print @ccc,"\n";
for $idx (reverse(@aaa)) {
	$bbb[$ccc[$idx]] = $idx;
	$ccc[$idx] -= 1;
}

for($idx=0;$idx<$dataset_size;$idx++){
	print $bbb[$idx],"\n";
}
